<div style="text-align:left;font-size:2em"><span style="font-weight:bolder;font-size:1.25em">SP2273 | Learning Portfolio</span><br><br><span style="font-weight:bold;color:darkred">Loops (Nice)</span></div>

# 1 There is more to list comprehension

In [1]:
# One can have more than one loop in a list comprehension
[[a, b] for a in range(5) for b in ['A', 'B', 'C']]

[[0, 'A'],
 [0, 'B'],
 [0, 'C'],
 [1, 'A'],
 [1, 'B'],
 [1, 'C'],
 [2, 'A'],
 [2, 'B'],
 [2, 'C'],
 [3, 'A'],
 [3, 'B'],
 [3, 'C'],
 [4, 'A'],
 [4, 'B'],
 [4, 'C']]

In [2]:
# One can also incorporate a condition.

[[a, b] for a in range(5) for b in ['A', 'B', 'C'] if a % 2 != 0]

[[1, 'A'], [1, 'B'], [1, 'C'], [3, 'A'], [3, 'B'], [3, 'C']]

In [6]:
# Below is a more complicated use of list comprehension to flatten a list.

nested_list = [[1, 2, 3], [4, 5, 6, 7]]
[y for x in nested_list for y in x]

# which does the same as the following:

nested_list = [[1, 2, 3], [4, 5, 6, 7]]
output = []
for x in nested_list:
    for y in x:
        output.append(y)
print(output)

[1, 2, 3, 4, 5, 6, 7]


# 2 Zipping a dictionary

In [7]:
# zip() offers one of the easiest ways to combine two lists into a dictionary:

super_names=["Black Widow", "Iron Man", "Doctor Strange"]
real_names=["Natasha Romanoff", "Tony Stark", "Stephen Strange"]

dict(zip(real_names, super_names))

# dict() is used to recast zip()'s output into a dictionary

{'Natasha Romanoff': 'Black Widow',
 'Tony Stark': 'Iron Man',
 'Stephen Strange': 'Doctor Strange'}

# 3 for and while has an else

In [8]:
# for and while loops come with an optional else statement.
# code in else block isn executed only if the loops are completed. 
# else code-block will not run if loop is exited prematurely (e.g. by using break)

numbers = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]

for i in numbers:
    if i < 0:
        break
else:
    print('No negative numbers in the list')

No negative numbers in the list


# Exercises

#### **Exercise 1 (Changing a List)**

In [13]:
fruits = ["apple", "banana", "jackfruit",
          "pineapple", "papaya", "watermelons",
          "peaches", "durian",  "mangoes",
          "strawberries", "passionfruit"
          ]

for fruit in fruits:
    if fruit[0] == "p":
        fruits.remove(fruit)
print(fruits)

# Doesn't work, why?
# Because during the For loop, when it reaches index [3] pineapple, the For loop will remove pineapple. The For loop will then look
# at the next item which is index [4]. However, while we expect 'papaya' to be in index [4], because the list has been changed,
# the 'papaya' has been shifted to index [3] because 'pineapple' has been removed, and therefore there is no chance for 'papaya'
# to be removed from the list as a result. It follows that the output will include 'papaya' which is incorrect.


# SOLUTION
fruits = ["apple", "banana", "jackfruit",
          "pineapple", "papaya", "watermelons",
          "peaches", "durian",  "mangoes",
          "strawberries", "passionfruit"
          ]

copy_of_fruits = fruits.copy()

# Not that we are looping(iterating) over the copy
for fruit in copy_of_fruits:
    if fruit[0] == "p":
        fruits.remove(fruit)
print(fruits)

# By including a copy, the loop is iterated through the copy, but the fruit itself is removed from the original list 'fruits'.
# Because the original list is not changed, therefore there will be no issues encountered in this For loop thus.


# MORE EFFICIENT SOLUTION
fruits = ["apple", "banana", "jackfruit",
          "pineapple", "papaya", "watermelons",
          "peaches", "durian",  "mangoes",
          "strawberries", "passionfruit"
          ]

[fruit for fruit in fruits if fruit[0] != "p"]

['apple', 'banana', 'jackfruit', 'papaya', 'watermelons', 'durian', 'mangoes', 'strawberries']
['apple', 'banana', 'jackfruit', 'watermelons', 'durian', 'mangoes', 'strawberries']


['apple',
 'banana',
 'jackfruit',
 'watermelons',
 'durian',
 'mangoes',
 'strawberries']

#### **Exercise 2 (A List of Powers)**

In [19]:
maximum_n = 5
result = [[]] * maximum_n

for n in range(1, maximum_n + 1):
    result[n - 1].append(n)
    result[n - 1].append(n**2)
    result[n - 1].append(n**3)

print(result)

# Problem is because [[]] * 5 creates 5 references to the same empty list. Thus
# 5 different empty lists will be created [].

# TO CHECK:
maximum_n = 5

result = [[]] * maximum_n

for count, element in enumerate(result):
    print(count, id(element))

maximum_n = 5

result = [[] for _ in range(maximum_n)]

for count, element in enumerate(result):
    print(count, id(element))

# SOLUTION
maximum_n = 5
# result = [[]] * maximum_n
result = [[] for _ in range(maximum_n)]

for n in range(1, maximum_n + 1):
    result[n - 1].append(n)
    result[n - 1].append(n**2)
    result[n - 1].append(n**3)

print(result)

[[1, 1, 1, 2, 4, 8, 3, 9, 27, 4, 16, 64, 5, 25, 125], [1, 1, 1, 2, 4, 8, 3, 9, 27, 4, 16, 64, 5, 25, 125], [1, 1, 1, 2, 4, 8, 3, 9, 27, 4, 16, 64, 5, 25, 125], [1, 1, 1, 2, 4, 8, 3, 9, 27, 4, 16, 64, 5, 25, 125], [1, 1, 1, 2, 4, 8, 3, 9, 27, 4, 16, 64, 5, 25, 125]]
0 2715490568960
1 2715490568960
2 2715490568960
3 2715490568960
4 2715490568960
0 2715490583872
1 2715490507392
2 2715490597440
3 2715490596352
4 2715490537920
[[1, 1, 1], [2, 4, 8], [3, 9, 27], [4, 16, 64], [5, 25, 125]]


#### **Exercise 3 (Time Profiling)**

In [5]:
%%timeit # %%timeit has to be the first thing in the cell

# Way 1: Creating a list of squares with a for loop
# 1 - Option 1 and 2 - Option 1

list = []

for i in range(1, 6):
    squared = i * i
    list.append(squared)
#print(list)

# Time Taken: 1.11 µs ± 14.1 ns per loop


1.11 µs ± 14.1 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [6]:
%%timeit

# Way 2: Creating a list of squares with While Loop
# 1 - Option 2

list = []
a = 1

while True:
    squared = a * a
    a += 1
    list.append(squared)
    if a > 5:
        break
#print(list)   # don't run %timeit with print

# Time Taken: 872 ns ± 24.3 ns per loop

872 ns ± 24.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [12]:
%%timeit

# Way 3: Creating a list of squares with a list comprehension loop.
# 2 - Option 2 and 5 - Option 2

[i*i for i in range(1, 6)]

# Time Taken: 1.29 µs ± 41.3 ns per loop

1.29 µs ± 41.3 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [3]:
%%timeit

# Way 4: Creating a list of squares using list.append()
# 3 - Option 1, 4 - Option 1

list = []

for i in range(1, 6):
    squared = i * i
    list.append(squared)
#print(list)

# Time Taken: 1.08 µs ± 43 ns per loop

1.08 µs ± 43 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [2]:
%%timeit

# Option 2: Creating a list of squares using list +=
# 3 - Option 2

list = []

for i in range(1, 6):
    list += [i*i]
#print(list)
    
# Time Taken: 1.6 µs ± 34.6 ns per loop

1.6 µs ± 34.6 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


In [6]:
%%timeit

# Option 1: Creating a List of Squares Using Numpy
# *Note there is no need for a for loop when making list of squares in numpy
# 5 - Option 1

import numpy as np
int_array = np.array(range(1, 6))
squared = int_array ** 2
#print(squared)

# Time Taken: 9.86 µs ± 304 ns per loop

9.86 µs ± 304 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [1]:
%%timeit

# Option 2: Creating a list of Squares using append() of Numpy
# 4 - Option 2

import numpy as np

int_array = np.array([])
list = []

for i in range(1, 6):
    squared = i*i
    int_array = np.append(int_array, squared)
    
#print(int_array)

# Time Taken: 

35.5 µs ± 818 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


**Comparing the Execution Speeds of Different Methods of Creating a List of Squares**

|#   |Option 1            |Option 2             |Faster Method |
|:--:|:--:                |:--:                 |:--:          |
|1   |`for` **Loop** <br> 1.11 µs ± 14.1 ns per loop| `while` **Loop** <br> 872 ns ± 24.3 ns per loop| **Option 2**|
|2|`for` **Loop** <br> 1.11 µs ± 14.1 ns per loop| **List Comprehension** <br> 1.29 µs ± 41.3 ns per loop| **Option 1**|
|3|**List** `append()` <br> 1.08 µs ± 43 ns per loop | **List** `+=` <br> 1.6 µs ± 34.6 ns per loop| **Option 1**|
|4|**List** `append()` <br> 1.08 µs ± 43 ns per loop | `append()` **on NumPy** <br> 35.5 µs ± 818 ns per loop| **Option 1** |
|5|**NumPy** <br> 9.86 µs ± 304 ns per loop | **List Comprehension** <br> 1.29 µs ± 41.3 ns per loop| **Option 2**|

Ranking:
1. `for` Loop
2. `while` Loop
3. List Comprehension
4. List `append()`
5. List `+=`
6. NumPy
7. `append()` on Numpy

#### **~ End of Loops (Nice) ~**